File: sanegor-backend/app/db/models/legal_source.py
"""The legal corpus: sources and their embedded chunks.

This is the only place citations may come from.  A :class:`LegalSource` is a
piece of primary material (a statute, a regulation, a ruling) that an operator
deliberately ingested; :class:`LegalChunk` is a retrievable slice of it with an
embedding.  The RAG pipeline can only cite chunk rows that exist here, which is
how the "never invent a reference" requirement is enforced structurally rather
than by asking the model nicely.
"""

from __future__ import annotations

import enum
import uuid
from datetime import date

from sqlalchemy import Date, ForeignKey, Index, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.config import get_settings
from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin
from app.db.types import VectorType, json_column

# The vector column width is fixed when the table is created, so it is read
# from configuration at import time rather than per-instance.
EMBEDDING_DIMENSIONS = get_settings().embedding_dimensions


class SourceType(str, enum.Enum):
    """Kind of primary material."""

    LEGISLATION = "legislation"      # חקיקה ראשית
    REGULATION = "regulation"        # תקנות
    RULING = "ruling"                # פסיקה
    GUIDELINE = "guideline"          # הנחיות רשות
    FORM = "form"                    # טפסים רשמיים


class CourtLevel(str, enum.Enum):
    """Israeli court hierarchy, for filtering rulings."""

    SUPREME = "supreme"              # בית המשפט העליון
    DISTRICT = "district"            # מחוזי
    MAGISTRATE = "magistrate"        # שלום
    LABOR_NATIONAL = "labor_national"  # ארצי לעבודה
    LABOR_REGIONAL = "labor_regional"  # אזורי לעבודה
    FAMILY = "family"                # לענייני משפחה
    TRAFFIC = "traffic"              # תעבורה
    ADMINISTRATIVE = "administrative"  # לעניינים מנהליים
    OTHER = "other"


class LegalDomain(str, enum.Enum):
    """Practice areas used to narrow retrieval."""

    CIVIL = "civil"
    CONTRACTS = "contracts"
    LABOR = "labor"
    FAMILY = "family"
    CRIMINAL = "criminal"
    ADMINISTRATIVE = "administrative"
    TENANCY = "tenancy"
    CONSUMER = "consumer"
    CORPORATE = "corporate"
    TORTS = "torts"
    PRIVACY = "privacy"
    TAX = "tax"
    OTHER = "other"


class LegalSource(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """A single ingested legal document."""

    __tablename__ = "legal_sources"
    __table_args__ = (
        UniqueConstraint("citation_key", name="uq_legal_sources_citation_key"),
        Index("ix_legal_sources_type_domain", "source_type", "domain"),
        Index("ix_legal_sources_published", "published_at"),
    )

    # Stable human-readable key, e.g. "חוק-החוזים-חלק-כללי-1973" — this is what
    # the model is shown and what the client resolves back to a source card.
    citation_key: Mapped[str] = mapped_column(String(160), nullable=False)
    title: Mapped[str] = mapped_column(String(400), nullable=False)
    short_title: Mapped[str | None] = mapped_column(String(200))
    source_type: Mapped[SourceType] = mapped_column(String(24), nullable=False, index=True)
    domain: Mapped[LegalDomain] = mapped_column(
        String(24), nullable=False, default=LegalDomain.OTHER, index=True
    )

    # Rulings
    case_number: Mapped[str | None] = mapped_column(String(80), index=True)
    court: Mapped[CourtLevel | None] = mapped_column(String(32), index=True)
    judges: Mapped[list] = mapped_column(json_column(), nullable=False, default=list)
    parties: Mapped[str | None] = mapped_column(String(400))
    proceeding_type: Mapped[str | None] = mapped_column(String(80), index=True)

    # Legislation
    section_range: Mapped[str | None] = mapped_column(String(80))
    amendment: Mapped[str | None] = mapped_column(String(120))

    published_at: Mapped[date | None] = mapped_column(Date)
    # Provenance is mandatory: an operator must record where material came from.
    source_url: Mapped[str | None] = mapped_column(String(1000))
    publisher: Mapped[str] = mapped_column(String(200), nullable=False, default="")
    language: Mapped[str] = mapped_column(String(8), nullable=False, default="he")
    checksum_sha256: Mapped[str | None] = mapped_column(String(64), index=True)
    chunk_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    extra: Mapped[dict] = mapped_column(json_column(), nullable=False, default=dict)

    chunks: Mapped[list[LegalChunk]] = relationship(
        back_populates="source", cascade="all, delete-orphan", lazy="raise"
    )

    def __repr__(self) -> str:  # pragma: no cover
        return f"<LegalSource {self.citation_key}>"


class LegalChunk(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """An embedded, retrievable slice of a :class:`LegalSource`.

    The embedding dimension is fixed at table-creation time from
    ``EMBEDDING_DIMENSIONS``; changing the embedding model therefore requires a
    migration, which is deliberate — silently mixing vector spaces would make
    retrieval quietly wrong.
    """

    __tablename__ = "legal_chunks"
    __table_args__ = (
        UniqueConstraint("source_id", "ordinal", name="uq_legal_chunks_source_ordinal"),
        Index("ix_legal_chunks_source", "source_id"),
    )

    source_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("legal_sources.id", ondelete="CASCADE")
    )
    ordinal: Mapped[int] = mapped_column(Integer, nullable=False)
    heading: Mapped[str | None] = mapped_column(String(300))
    content: Mapped[str] = mapped_column(Text, nullable=False)
    token_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    embedding: Mapped[list[float] | None] = mapped_column(
        VectorType(EMBEDDING_DIMENSIONS), nullable=True
    )

    source: Mapped[LegalSource] = relationship(back_populates="chunks", lazy="joined")

File: sanegor-app/lib/features/search/domain/legal_source.dart
import 'package:flutter/foundation.dart';

/// A record in the legal corpus.
@immutable
class LegalSource {
  const LegalSource({
    required this.id,
    required this.citationKey,
    required this.title,
    required this.sourceType,
    required this.domain,
    this.shortTitle,
    this.caseNumber,
    this.court,
    this.judges = const [],
    this.parties,
    this.proceedingType,
    this.publishedAt,
    this.sourceUrl,
    this.publisher = '',
    this.chunkCount = 0,
  });

  final String id;
  final String citationKey;
  final String title;
  final String sourceType;
  final String domain;
  final String? shortTitle;
  final String? caseNumber;
  final String? court;
  final List<String> judges;
  final String? parties;
  final String? proceedingType;
  final DateTime? publishedAt;
  final String? sourceUrl;
  final String publisher;
  final int chunkCount;

  factory LegalSource.fromJson(Map<String, dynamic> json) => LegalSource(
        id: (json['id'] ?? '').toString(),
        citationKey: (json['citation_key'] ?? '').toString(),
        title: (json['title'] ?? '').toString(),
        sourceType: (json['source_type'] ?? 'other').toString(),
        domain: (json['domain'] ?? 'other').toString(),
        shortTitle: json['short_title']?.toString(),
        caseNumber: json['case_number']?.toString(),
        court: json['court']?.toString(),
        judges: (json['judges'] as List? ?? const [])
            .map((e) => e.toString())
            .toList(growable: false),
        parties: json['parties']?.toString(),
        proceedingType: json['proceeding_type']?.toString(),
        publishedAt: DateTime.tryParse((json['published_at'] ?? '').toString()),
        sourceUrl: json['source_url']?.toString(),
        publisher: (json['publisher'] ?? '').toString(),
        chunkCount: (json['chunk_count'] as num?)?.toInt() ?? 0,
      );
}

/// A matched passage inside a source.
@immutable
class LegalPassage {
  const LegalPassage({
    required this.source,
    required this.snippet,
    required this.score,
    this.heading,
  });

  final LegalSource source;
  final String snippet;
  final double score;
  final String? heading;

  factory LegalPassage.fromJson(Map<String, dynamic> json) => LegalPassage(
        source: LegalSource.fromJson(
          Map<String, dynamic>.from(json['source'] as Map? ?? const {}),
        ),
        snippet: (json['snippet'] ?? '').toString(),
        score: (json['score'] as num?)?.toDouble() ?? 0,
        heading: json['heading']?.toString(),
      );
}

/// Result of a corpus search.
///
/// [corpusEmpty] distinguishes "nothing has been loaded into this deployment"
/// from "no match" — a distinction the user needs, because the first is a
/// configuration state and the second is a legal one.
@immutable
class LegalSearchResult {
  const LegalSearchResult({
    this.sources = const [],
    this.passages = const [],
    this.total = 0,
    this.corpusEmpty = false,
    this.notice,
  });

  final List<LegalSource> sources;
  final List<LegalPassage> passages;
  final int total;
  final bool corpusEmpty;
  final String? notice;

  bool get isEmpty => sources.isEmpty && passages.isEmpty;

  factory LegalSearchResult.fromJson(Map<String, dynamic> json) =>
      LegalSearchResult(
        sources: (json['sources'] as List? ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(LegalSource.fromJson)
            .toList(growable: false),
        passages: (json['passages'] as List? ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(LegalPassage.fromJson)
            .toList(growable: false),
        total: (json['total'] as num?)?.toInt() ?? 0,
        corpusEmpty: json['corpus_empty'] == true,
        notice: json['notice']?.toString(),
      );
}

/// Search filters mirroring the backend's request model.
@immutable
class SearchFilters {
  const SearchFilters({
    this.sourceTypes = const [],
    this.domains = const [],
    this.courts = const [],
    this.proceedingType,
    this.dateFrom,
    this.dateTo,
  });

  final List<String> sourceTypes;
  final List<String> domains;
  final List<String> courts;
  final String? proceedingType;
  final DateTime? dateFrom;
  final DateTime? dateTo;

  bool get isEmpty =>
      sourceTypes.isEmpty &&
      domains.isEmpty &&
      courts.isEmpty &&
      proceedingType == null &&
      dateFrom == null &&
      dateTo == null;

  int get activeCount =>
      sourceTypes.length +
      domains.length +
      courts.length +
      (proceedingType != null ? 1 : 0) +
      (dateFrom != null ? 1 : 0) +
      (dateTo != null ? 1 : 0);

  Map<String, dynamic> toJson() => {
        if (sourceTypes.isNotEmpty) 'source_types': sourceTypes,
        if (domains.isNotEmpty) 'domains': domains,
        if (courts.isNotEmpty) 'courts': courts,
        if (proceedingType != null) 'proceeding_type': proceedingType,
        if (dateFrom != null)
          'date_from': dateFrom!.toIso8601String().split('T').first,
        if (dateTo != null) 'date_to': dateTo!.toIso8601String().split('T').first,
      };

  SearchFilters copyWith({
    List<String>? sourceTypes,
    List<String>? domains,
    List<String>? courts,
    String? proceedingType,
    DateTime? dateFrom,
    DateTime? dateTo,
    bool clearDates = false,
  }) =>
      SearchFilters(
        sourceTypes: sourceTypes ?? this.sourceTypes,
        domains: domains ?? this.domains,
        courts: courts ?? this.courts,
        proceedingType: proceedingType ?? this.proceedingType,
        dateFrom: clearDates ? null : (dateFrom ?? this.dateFrom),
        dateTo: clearDates ? null : (dateTo ?? this.dateTo),
      );
}

/// Labels for the filter chips.
abstract final class LegalTaxonomy {
  static const sourceTypes = {
    'legislation': 'חקיקה',
    'regulation': 'תקנות',
    'ruling': 'פסיקה',
    'guideline': 'הנחיות',
    'form': 'טפסים',
  };

  static const courts = {
    'supreme': 'עליון',
    'district': 'מחוזי',
    'magistrate': 'שלום',
    'labor_national': 'ארצי לעבודה',
    'labor_regional': 'אזורי לעבודה',
    'family': 'משפחה',
    'traffic': 'תעבורה',
    'administrative': 'מנהליים',
  };

  static const domains = {
    'civil': 'אזרחי',
    'contracts': 'חוזים',
    'labor': 'עבודה',
    'family': 'משפחה',
    'criminal': 'פלילי',
    'administrative': 'מנהלי',
    'tenancy': 'שכירות',
    'consumer': 'צרכנות',
    'corporate': 'תאגידים',
    'torts': 'נזיקין',
    'privacy': 'פרטיות',
    'tax': 'מיסים',
  };
}
